File: LeetCode/Day 4/Solution.py
def is_anagram(s, t):
    """Determines if two strings are anagrams."""

    # Handle cases where lengths are different or strings contain non-letter characters:
    if len(s) != len(t) or not s.isalpha() or not t.isalpha():
        return False

    # Create a dictionary to count letter frequencies in the first string:
    char_counts = {}
    for char in s:
        char_counts[char] = char_counts.get(char, 0) + 1

    # Decrement counts for letters in the second string:
    for char in t:
        if char not in char_counts or char_counts[char] == 0:
            return False
        char_counts[char] -= 1

    # If all counts are 0, the strings are anagrams:
    return all(count == 0 for count in char_counts.values())


print(is_anagram("anagram", "nagaram"))  # Output: True
print(is_anagram("rat", "car"))         # Output: False
print(is_anagram("listen", "silent"))   # Output: True

File: LeetCode/Day 4/Solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

bool is_anagram(const string& s, const string& t) {
    if (s.length() != t.length()) {
        return false;
    }

    // Check if both strings contain only letters using a loop:
    for (char c : s) {
        if (!isalpha(c)) {
            return false;
        }
    }
    for (char c : t) {
        if (!isalpha(c)) {
            return false;
        }
    }

    unordered_map<char, int> char_counts;
    for (char c : s) {
        ++char_counts[c];
    }

    for (char c : t) {
        if (--char_counts[c] < 0) {
            return false;
        }
    }

    return true;
}

int main() {
    string s1 = "anagram";
    string s2 = "nagaram";
    string s3 = "rat";
    string s4 = "car";

    cout << s1 << " and " << s2 << " are anagrams? " << (is_anagram(s1, s2) ? "Yes" : "No") << endl;
    cout << s3 << " and " << s4 << " are anagrams? " << (is_anagram(s3, s4) ? "Yes" : "No") << endl;

    return 0;
}
